diff --git a/android/README.md b/android/README.md
deleted file mode 100644
index 6eebb68d9f242a118d634455976c94c3936dd56c..0000000000000000000000000000000000000000
diff --git a/ios/Core/CameraSession.swift b/ios/Core/CameraSession.swift
index 10b0f3399c220fd1ab6de286bba6a5b641d5bf30..0e1d6fd4154a996fd3f1805c4225220efa3a495d 100644
--- a/ios/Core/CameraSession.swift
+++ b/ios/Core/CameraSession.swift
@@ -113,8 +113,22 @@ final class CameraSession: NSObject, AVCaptureVideoDataOutputSampleBufferDelegat
 
     VisionLogger.log(level: .info, message: "configure { ... }: Waiting for lock...")
 
+    let slowConfigurationWarning = DispatchWorkItem {
+      VisionLogger.log(level: .warning, message: "configure { ... }: is still running after 2 seconds.")
+    }
+    CameraQueues.cameraQueue.asyncAfter(deadline: .now() + .seconds(2), execute: slowConfigurationWarning)
+
+    let completionGroup = DispatchGroup()
+
+    completionGroup.enter()
+    completionGroup.notify(queue: CameraQueues.cameraQueue) {
+      slowConfigurationWarning.cancel()
+      VisionLogger.log(level: .info, message: "configure { ... }: completed.")
+    }
+
     // Set up Camera (Video) Capture Session (on camera queue, acts like a lock)
     CameraQueues.cameraQueue.async {
+      defer { completionGroup.leave() }
       // Let caller configure a new configuration for the Camera.
       let config = CameraConfiguration(copyOf: self.configuration)
       do {
@@ -215,7 +229,9 @@ final class CameraSession: NSObject, AVCaptureVideoDataOutputSampleBufferDelegat
 
       // Set up Audio Capture Session (on audio queue)
       if difference.audioSessionChanged {
+        completionGroup.enter()
         CameraQueues.audioQueue.async {
+          defer { completionGroup.leave() }
           do {
             // Lock Capture Session for configuration
             VisionLogger.log(level: .info, message: "Beginning AudioSession configuration...")
@@ -234,7 +250,9 @@ final class CameraSession: NSObject, AVCaptureVideoDataOutputSampleBufferDelegat
 
       // Set up Location streaming (on location queue)
       if difference.locationChanged {
+        completionGroup.enter()
         CameraQueues.locationQueue.async {
+          defer { completionGroup.leave() }
           do {
             VisionLogger.log(level: .info, message: "Beginning Location Output configuration...")
             try self.configureLocationOutput(configuration: config)
@@ -265,7 +283,7 @@ final class CameraSession: NSObject, AVCaptureVideoDataOutputSampleBufferDelegat
     }
   }
 
-  public final func captureOutput(_ captureOutput: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+  final func captureOutput(_ captureOutput: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
     switch captureOutput {
     case is AVCaptureVideoDataOutput:
       onVideoFrame(sampleBuffer: sampleBuffer, orientation: connection.orientation, isMirrored: connection.isVideoMirrored)
diff --git a/ios/Core/Utils/Counter.swift b/ios/Core/Utils/Counter.swift
new file mode 100644
index 0000000000000000000000000000000000000000..b2d4378a19a07eace8ce52040808892fbe8da357
--- /dev/null
+++ b/ios/Core/Utils/Counter.swift
@@ -0,0 +1,34 @@
+import os.lock
+
+final class Counter {
+  /**
+    * https://forums.swift.org/t/atomic-property-wrapper-for-standard-library/30468/18
+  */
+  private var unfair_lock: os_unfair_lock_t
+  private var count = 1
+  init() {
+    unfair_lock = .allocate(capacity: 1)
+    unfair_lock.initialize(to: os_unfair_lock())
+  }
+  deinit {
+    unfair_lock.deinitialize(count: 1)
+    unfair_lock.deallocate()
+  }
+  private func lock() {
+    os_unfair_lock_lock(unfair_lock)
+  }
+  private func unlock() {
+    os_unfair_lock_unlock(unfair_lock)
+  }
+  func increment() -> Int {
+    lock()
+    defer { unlock() }
+    count &+= 1
+    return count
+  }
+  func check(_ count: Int) -> Bool {
+    lock()
+    defer { unlock() }
+    return self.count == count
+  }
+}
\ No newline at end of file
diff --git a/ios/README.md b/ios/README.md
deleted file mode 100644
index 1c47fe132d3ab5ef1ed73fdc309ce34b91cf4c31..0000000000000000000000000000000000000000
diff --git a/ios/React/CameraView.swift b/ios/React/CameraView.swift
index c773975353865c548e6ae7aef63d8575b6c395ad..58cd1365be96caad692196e0762f8ff85f4fdbad 100644
--- a/ios/React/CameraView.swift
+++ b/ios/React/CameraView.swift
@@ -10,6 +10,8 @@ import AVFoundation
 import Foundation
 import UIKit
 
+import os.lock
+
 // TODOs for the CameraView which are currently too hard to implement either because of AVFoundation's limitations, or my brain capacity
 //
 // CameraView+RecordVideo
@@ -101,7 +103,7 @@ public final class CameraView: UIView, CameraSessionDelegate, PreviewViewDelegat
   var cameraSession = CameraSession()
   var previewView: PreviewView?
   var isMounted = false
-  private var currentConfigureCall: DispatchTime?
+  private let currentConfigureCall: Counter = .init()
   private let fpsSampleCollector = FpsSampleCollector()
 
   // CameraView+Zoom
@@ -135,6 +137,7 @@ public final class CameraView: UIView, CameraSessionDelegate, PreviewViewDelegat
         onViewReadyEvent?(nil)
       }
     } else {
+      deactivateCameraSession()
       fpsSampleCollector.stop()
     }
   }
@@ -181,12 +184,12 @@ public final class CameraView: UIView, CameraSessionDelegate, PreviewViewDelegat
   // pragma MARK: Props updating
   override public final func didSetProps(_ changedProps: [String]!) {
     VisionLogger.log(level: .info, message: "Updating \(changedProps.count) props: [\(changedProps.joined(separator: ", "))]")
-    let now = DispatchTime.now()
-    currentConfigureCall = now
+    let currentConfigureCall = self.currentConfigureCall
+    let now = currentConfigureCall.increment()
 
     cameraSession.configure { [self] config in
       // Check if we're still the latest call to configure { ... }
-      guard currentConfigureCall == now else {
+      guard currentConfigureCall.check(now) else {
         // configure waits for a lock, and if a new call to update() happens in the meantime we can drop this one.
         // this works similar to how React implemented concurrent rendering, the newer call to update() has higher priority.
         VisionLogger.log(level: .info, message: "A new configure { ... } call arrived, aborting this one...")
@@ -283,6 +286,33 @@ public final class CameraView: UIView, CameraSessionDelegate, PreviewViewDelegat
     UIApplication.shared.isIdleTimerDisabled = isActive
   }
 
+    private func deactivateCameraSession() {
+    UIApplication.shared.isIdleTimerDisabled = false
+
+    #if VISION_CAMERA_ENABLE_FRAME_PROCESSORS
+      frameProcessor = nil
+    #endif
+
+    let currentConfigureCall = self.currentConfigureCall
+    let now = currentConfigureCall.increment()
+
+    cameraSession.configure { config in
+      guard currentConfigureCall.check(now) else {
+        VisionLogger.log(level: .info, message: "A new configure { ... } call arrived, aborting this one [\(now)]…")
+        throw CameraConfiguration.AbortThrow.abort
+      }
+      VisionLogger.log(level: .info, message: "configure { ... } [\(now)]")
+      config.photo = .disabled
+      config.video = .disabled
+      config.audio = .disabled
+      config.codeScanner = .disabled
+      config.enableLocation = false
+      config.torch = .off
+      config.isActive = false
+    }
+  }
+
+
   func updatePreview() {
     if preview && previewView == nil {
       // Create PreviewView and add it
